airflow/hooks/fs_hooks/base.py

```diff
@@ -0,0 +1,231 @@
+        on its connection type.
+
+        :param str conn_type: Connection type.
+        :param class_: FsHook to register. Can either be the class itself
+            or a string specifying the full module path for the class.
+        """
+        cls._conn_classes[conn_type] = class_
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.disconnect()
+
+    def disconnect(self):
+        """Closes fs connection (if applicable)."""
+        pass
+
+    # Interface methods (should be implemented by sub-classes).
+
+    def open(self, file_path, mode='rb'):
+        """Returns file_obj for given file path.
+
+        :param str file_path: Path to the file to open.
+        :param str mode: Mode to open the file in.
+
+        :returns: An opened file object.
+        """
+        raise NotImplementedError()
+
+    def exists(self, file_path):
+        """Checks whether the given file path exists.
+
+        :param str file_path: File path.
+
+        :returns: True if the file exists, else False.
+        :rtype: bool
+        """
+        raise NotImplementedError()
+
+    def isdir(self, path):
+        """Returns true if the given path points to a directory.
+
+        :param str path: File or directory path.
+        """
+        raise NotImplementedError()
+
+    def walk(self, dir_path):
+        """Generates file names in the given directory tree."""
+        raise NotImplementedError()
+
+    def makedir(self, dir_path, mode=0o755, exist_ok=True):
+        """Creates the directory, without creating intermediate directories."""
+        raise NotImplementedError()
+
+    def makedirs(self, dir_path, mode=0o755, exist_ok=True):
+        """Creates directory, creating intermediate directories if needed.
+
+        :param str dir_path: Path to the directory to create.
+        :param int mode: Mode to use for directory (if created).
+        :param bool exist_ok: Whether the directory is already allowed to exist.
+            If false, an IOError is raised if the directory exists.
+        """
+        raise NotImplementedError()
+
+    def rm(self, file_path):
+        """Deletes the given file path.
+
+        :param str file_path: Path to file:
+        """
+        raise NotImplementedError()
+
+    def rmtree(self, dir_path):
+        """Deletes given directory tree recursively.
+
+        :param str dir_path: Path to directory to delete.
+        """
+        raise NotImplementedError()
+
+    @staticmethod
+    def _raise_dir_exists(dir_path):
+        raise IOError(errno.EEXIST,
+                      'Directory exists: {!r}'.format(dir_path))
+
+    # Path manipulation methods.
+
+    # Join/split should be overriden for file systems that don't use
+    # posix path conventions.
+
+    @staticmethod
+    def join(path, *paths):
+        """Join one or more path components."""
+        return posixpath.join(path, *paths)
```

Comment:
Do these methods really add any value over just calling `posixpath.join` directly? The smaller the interface the easier it is to reason about.

Reply:
I agree that a smaller interface is better, however I wanted to keep the flexibility for supporting file systems that do not necessarily use posix-path style joins. If this is only different on windows, we can consider just using posixpath. (Is Airflow supposed to support windows?)

Reply:
We could also keep the methods internal and default to posixpath, allowing subclasses to override if necessary.

Reply:
All versions of windows since 7? Vista? have supported `/` in paths - i.e. `c:/foo.txt` works so we don't need to worry about this and can remove it.